exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/RemoteFunctionRegistry.java

```diff
@@ -0,0 +1,185 @@
+
+/** Is responsible for remote function registry management.
+ *  Creates all remote registry areas at startup, during init establishes connections with stores.
+ *  Provides tools to update remote registry and unregister functions, access remote registry areas.
+ */
+public class RemoteFunctionRegistry implements AutoCloseable {
+
+  public static final String REGISTRY = "registry";
+
+  private static final ObjectMapper mapper = new ObjectMapper().enable(INDENT_OUTPUT);
+
+  private final TransientStoreListener unregistrationListener;
+  private int retryTimes;
+  private FileSystem fs;
+  private Path registryArea;
+  private Path stagingArea;
+  private Path tmpArea;
+
+  private PersistentStore<Registry> registry;
+  private TransientStore<String> unregistration;
+  private TransientStore<String> jars;
+
+  public RemoteFunctionRegistry(TransientStoreListener unregistrationListener) {
+    this.unregistrationListener = unregistrationListener;
+  }
+
+  public void init(DrillConfig config, PersistentStoreProvider storeProvider, ClusterCoordinator coordinator) {
+    try {
+      PersistentStoreConfig<Registry> registrationConfig = PersistentStoreConfig
+          .newProtoBuilder(SchemaUserBitShared.Registry.WRITE, SchemaUserBitShared.Registry.MERGE)
+          .name("udf")
+          .persist()
+          .build();
+      registry = storeProvider.getOrCreateStore(registrationConfig);
+      registry.putIfAbsent(REGISTRY, Registry.getDefaultInstance());
+    } catch (StoreException e) {
+      throw new DrillRuntimeException("Failure while loading remote registry.", e);
+    }
+
+    TransientStoreConfig<String> unregistrationConfig = TransientStoreConfig.
+        newJacksonBuilder(mapper, String.class).name("udf/unregister").build();
+    unregistration = coordinator.getOrCreateTransientStore(unregistrationConfig);
+    unregistration.addListener(unregistrationListener);
+
+    TransientStoreConfig<String> jarsConfig = TransientStoreConfig.
+        newJacksonBuilder(mapper, String.class).name("udf/jars").build();
+    jars = coordinator.getOrCreateTransientStore(jarsConfig);
+
+    this.retryTimes = config.getInt(ExecConstants.UDF_RETRY_TIMES);
+
+    Configuration conf = new Configuration();
+    if (config.hasPath(ExecConstants.UDF_DIRECTORY_FS)) {
+      conf.set(FileSystem.FS_DEFAULT_NAME_KEY, config.getString(ExecConstants.UDF_DIRECTORY_FS));
+    }
+
+    try {
+      this.fs = FileSystem.get(conf);
+      String base = config.getString(ExecConstants.UDF_DIRECTORY_BASE);
+      this.registryArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_REGISTRY));
+      this.stagingArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_STAGING));
+      this.tmpArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_TMP));
+    } catch (IOException e) {
+      throw new DrillRuntimeException("Error during areas creation for remote function registry.", e);
```

Comment:
The most likely place that stuff will break is in the external interfaces. I'd suggest modifying your createArea method to take more information such as the name of the property. Then, if the creation of the directory fails, you can issue a very specific exception that points the user to the property that is being used, the value of that property, and the specific reason. Such as "Cannot create directory /a/b/c on file system foo://bar as specified in property drill.exec.mumble."